src/main/java/fi/pohina/vinkkilista/App.java

```diff
@@ -31,45 +67,182 @@ public void startServer(int portNumber) {
-                return "New bookmark added";
-            } else {
-                res.redirect("/new");
-                return "Bookmark could not be created";
-            }
-        });
+                return render(map, "index");
+            });
 
-        post("/search", (req, res) -> {
-            Map<String, Object> map = new HashMap<>();
-            Collection<Bookmark> bookmarks = searchBookmarks(req.queryMap());
-            map.put("bookmarks", bookmarks);
-            return render(map, "search");
+            //Turha
+            get("/bookmarks", (req, res) -> {
+                Map<String, Object> map = new HashMap<>();
+                Collection<Bookmark> bookmarks = this.bookmarks.getAllBookmarks();
+                map.put("bookmarks", bookmarks);
+                return render(map, "index");
+            });
+
+            get("/new", (req, res) -> {
+                Map<String, Object> map = new HashMap<>();
+                return render(map, "new");
+            });
+
+            get("/search", (req, res) -> {
+                Map<String, Object> map = new HashMap<>();
+                return render(map, "search");
+            });
+
+            post("/new", (req, res) -> {
+                if (validateAndCreateBookmark(req.queryMap())) {
+                    res.redirect("/bookmarks/");
+                    return "New bookmark added";
+                } else {
+                    res.redirect("new");
+                    return "Bookmark could not be created";
+                }
+            });
+
+            post("/search", (req, res) -> {
+                Map<String, Object> map = new HashMap<>();
+                Collection<Bookmark> bookmarks = searchBookmarks(req.queryMap());
+                map.put("bookmarks", bookmarks);
+                return render(map, "search");
+            });
         });
 
+
         get("/login", (req, res) -> {
             Map<String, Object> map = new HashMap<>();
             map.put("clientId", config.getGithubClientId());
             return render(map, "login");
         });
+
+        get("/auth/gh-callback", (req, res) -> {
+            String callbackCode = req.queryParams("code");
+
+            String accessToken = postAuthCallbackCode(callbackCode);
+
+            // 2. get user info
+            String userInfoJson = getUserInfoJson(accessToken);
+
+            GithubUser githubUser = new Gson().fromJson(userInfoJson, GithubUser.class);
+
+            // 3. get user emails because if user has multiple emails, we need to find the primary one
+            if (githubUser.getEmail() == null) {
+                githubUser.setEmail(getPrimaryUserEmail(accessToken));
+            }
+
+            User user = users.findOrCreateByGithubUser(githubUser);
+
+            req.session(true)
+                    .attribute(SESSION_ATTRIBUTE_USERID, user.getId());
+
+            res.redirect("/bookmarks/");
+            return "";
+        });
+    }
+
```

Comment:
Tää luokka rupee olee aika härski, mun mielestä pitäis siirtää kaikki muu paitsi routet pois täältä